R/use-bbi.R

```diff
@@ -62,57 +86,94 @@ current_release <- function(owner = 'metrumresearchgroup', repo = 'babylon', os
+  str_replace(basename(dirname(current_release(owner = 'metrumresearchgroup', repo = 'babylon', os = os))), '^v', '')
 }
 
-install_menu <- function(body,this_os){
-
-  release_v <- bbi_current_release()
-  local_v <- bbi_version()
+#' Private implementation function for installing bbi with interactive menu
+#' @param .body Character vector of installation commands to run with `system`
+#' @param .this_os Character scaler of OS
+#' @param .dir directory to install bbi to
+#' @param .version version of babylon to install. Must pass a character scaler corresponding to a tag found in `https://github.com/metrumresearchgroup/babylon/releases`
+#' @param .force Boolean for whether to force the installation even if current version and local version are the same
+#' @param .quiet Boolean for suppressing output printed to the console. False by default.
+install_menu <- function(.body, .this_os, .dir, .version, .force, .quiet){
+
+  .dest_bbi_path <- normalizePath(file.path(.dir, "bbi"), mustWork = FALSE)
+  local_v <- bbi_version(.dest_bbi_path)
+
+  if (.version == 'latest') {
+    release_v <- bbi_current_release()
+  } else {
+    release_v <- .version
+  }
 
-  if(!identical(release_v,local_v)){
+  if(!identical(release_v,local_v) || isTRUE(.force)){
 
-    if(this_os%in%c('linux','darwin')){
+    if(.this_os%in%c('linux','darwin')){
 
       # suppressing interactivity will allow for suppression in unit tests. This may be more general to name
       # like specifiying its a test environment, but a user could also want to generally suppress interactivity for
       # automatted build pipelines etc.
       if (interactive() && is.null(getOption('rbabylon.suppress_interactivity'))) {
-        print(glue::glue(cli::rule(left = cli::col_red('{local_v} is currently installed. Update to the {release_v}?'),line = 2)))
+        version_message(local_v = local_v, release_v = release_v)
+
+        print(glue::glue(cli::rule(left = cli::col_red('Do you want to install version {release_v}?'),line = 2)))
 
         if(utils::menu(choices = c('Yes','No'))==1){
-          system(paste(body,collapse =' ; '))
-          local_v <- bbi_version()
+          map(.body, ~ system(.x, ignore.stdout = .quiet, ignore.stderr = .quiet))
+          local_v <- bbi_version(.dest_bbi_path)
         }
       } else {
-          system(paste(body,collapse =' ; '))
+        map(.body, ~ system(.x, ignore.stdout = .quiet, ignore.stderr = .quiet))
       }
 
     }
 
   }
 
-  version_message(local_v = local_v,release_v = release_v)
+  version_message(local_v = local_v, release_v = release_v)
+}
+
+
+#' Private helper function for building commands to install bbi on Linux
+#' @param .dir Directory to install into
+#' @param .bbi_url Full url to download tarball from
+linux_install_commands <- function(.dir, .bbi_url) {
+
+  bbi_loc <- normalizePath(file.path(.dir, "bbi"), mustWork = FALSE)
+
+  cmd_vec <- c(
+    glue::glue('wget {.bbi_url} -O /tmp/bbi.tar.gz  --timeout=15 --tries=2'),
```

Comment:
I realize this was refactored from `use_bbi()`, and that the design of `use_bbi()` assumes that we will pass a vector of system commands to `install_menu()`, but it's worth considering another pattern. Specifically, downloading files and performing local filesystem operations can be done with R functions (versus system calls), for example `download.file()`. Given that we want to use a menu interactively, we could also do something like
```r
confirm_install <- function(.this_os) {
  if (interactive() {
    # interactive menu stuff
    # if user does not confirm installation
    stop("installation halted")
  }
  invisible(TRUE)
}

install_bbi <- function(.this_os, .version) {
  UseMethod("install_bbi")
}

install_bbi.linux <- function(.this_os, .version) {
  bbi_url <- construct_bbi_url(.version)
  tmp <- tempfile()
  download.file(bbi_url, destfile = tmp)
  # filesystem operations to install bbi
}

install_bbi.darwin <- function(.this_os, .version) {
  # macOS installation, possibly system calls to brew
}

use_bbi <- function(.version) {
  .this_os <- get_os()
  confirm_install(.this_os)
  install_bbi(.this_os, .version)
}
```
The benefit of this design is that all OS-specific behavior is encapsulated by the corresponding methods (and maybe we do something a bit more sophisticated around checking inputs). It also frees us from a single paradigm for installing babylon (a vector of system commands), which means we can use R functions for Linux and (if necessary) system calls for macOS.